### PR TITLE
Implement Persistent Subscription to $all

### LINKED ISSUE #96: Align Persistent Subscription naming and behaviour with RFC-022 - gRPC Persistent Subscriptions
As the next release will be breaking, then we could use it to align the naming and behaviour with what was agreed in the RFC-022 - gRPC Persistent Subscriptions. Current parameters names are not aligned.

[RFC-022](https://github.com/EventStore/architecture-and-planning/blob/main/rfcs/022%20-%20gRPC%20Persistent%20Subscriptions.md)

[Original Issue](https://github.com/EventStore/architecture-and-planning/issues/42)

We can also consider doing that non-breaking by marking the old names as obsolete and passing the getters and setters to the new options.

### PR DESCRIPTION
**Pull request by YoEight**
Fixes #96

### CODE REVIEW

**Comment by YoEight** [comment]
Still needs tests

**Comment by YoEight** [comment]
Need to find a way to handle not supported persistent subscriptions to $all operation on servers <21.*

**Comment by ylorph** [comment]
in examples/subscribing_to_stream.rs line 106

`pub async fn subscribe_to_filtered(client: &Client) -> Result<()> {`
would better be called 
`pub async fn subscribe_to_all_filtered(client: &Client) -> Result<()> {`

File: src/client.rs [hunk 1]
```diff
@@ -362,6 +368,19 @@ impl Client {
         .await
     }
 
+    /// Creates a persistent subscription group on a the $all stream.
+    pub async fn create_persistent_subscription_to_all<GroupName>(
+        &self,
+        group_name: GroupName,
+        options: &PersistentSubscriptionToAllOptions,
```
Comment (ylorph) [inline src/client.rs:375]: What's the difference with the PersistentSubscrptionOption ?
int the .net Client the options for create & createToAll are the same , not a different structure

Comment (ylorph) [reply src/client.rs:375]: Ah ok I see , it

**Comment by ylorph** [comment]
Would be nice to have a set of samples like 
https://github.com/EventStore/EventStore-Client-Dotnet/blob/master/samples/persistent-subscriptions/Program.cs#L52
so taht we can integrate in the docs later on

**Review by ylorph — COMMENTED** [review]
Would be nice to have a set of samples like
https://github.com/EventStore/EventStore-Client-Dotnet/blob/master/samples/persistent-subscriptions/Program.cs#L52
so taht we can integrate in the docs later on

File: examples/persistent_subscriptions.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+async fn connect_to_persistent_subscription_to_stream(client: &Client) -> eventstore::Result<()> {
+    // #region subscribe-to-persistent-subscription-to-stream
+    let (mut read, mut write) = client
+        .connect_persistent_subscription("test-stream", "subscription-group", &Default::default())
+        .await?;
+
+    while let Some(event) = read.try_next_event().await? {
+        // Doing some productive work with the event...
+        let _ = write.ack_event(event).await;
```
Comment (ylorph) [inline examples/persistent_subscriptions.rs:24]: We have a discrepency in the informations we get back in this method between at leas .net & rust (RFC022 needs some addedum to make it clear ) 
.net give the Subscription client, retry count & resolved event.

**Review by ylorph — CHANGES_REQUESTED** [review]

File: examples/persistent_subscriptions.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    Ok(())
+}
+
+async fn create_persistent_subscription_to_all(client: &Client) -> eventstore::Result<()> {
+    // #region create-persistent-subscription
+    let options = PersistentSubscriptionToAllOptions::default()
+        .filter(SubscriptionFilter::on_stream_name().add_prefix("test"));
+
+    client
+        .create_persistent_subscription_to_all("subscription-group", &options)
```
Comment (ylorph) [inline examples/persistent_subscriptions.rs:36]: Idem above, rFC022: subscribe_to_all

**Review by ylorph — CHANGES_REQUESTED** [review]

File: examples/persistent_subscriptions.rs (hunk 2 above)
Comment (YoEight) [reply examples/persistent_subscriptions.rs:24]: Done

File: examples/persistent_subscriptions.rs (hunk 3 above)
Comment (ylorph) [reply examples/persistent_subscriptions.rs:45]: ah, no this is effectively the create action, not subscribing to it

File: src/client.rs [hunk 4]
```diff
@@ -418,7 +476,243 @@ impl Client {
             stream_name.as_ref(),
             group_name.as_ref(),
             options,
+            false,
+        )
+        .await
+    }
+
+    /// Connects to a persistent subscription group to $all stream.
+    pub async fn connect_persistent_subscription_to_all<GroupName>(
```
Comment (ylorph) [inline src/client.rs:485]: so we need to decide on the naming convention as per RFC022

Comment (ylorph) [reply src/client.rs:485]: if we don't want to wait too long, we could create another issue for this ?
I guess in the Rust world , this is not too bad to rename things , people are used to it ?

Comment (YoEight) [reply src/client.rs:485]: I opened the issue there: #108

**Review by ylorph — APPROVED** [review]

**Comment by YoEight** [comment]
I renamed all persistent subscriptions functions to new RFC 22 names.

**Review by ylorph — APPROVED** [review]

File: src/options/subscribe_to_stream.rs [hunk 5]
```diff
@@ -35,7 +35,7 @@ impl SubscribeToStreamOptions {
     /// as the subscription is dropped or closed.
     ///
     /// By default, it's `Origin::Start`
```
[the remark this answers is not in the document]
Comment (YoEight) [reply src/options/subscribe_to_stream.rs]: done

**Review by ylorph — APPROVED** [review]

**Review by arwinneil — APPROVED** [review]

**Pull request merged**